File: src/audioWfc/wfc/constraints/ChordPrototypeOnlyPrecededByConstraint.java
package audioWfc.wfc.constraints;

import audioWfc.musicTheory.chords.Chord;
import audioWfc.wfc.HigherValues;
import audioWfc.wfc.Tile;
import audioWfc.wfc.constraints.concepts.HardConstraint;
import audioWfc.wfc.grabbers.Grabber;
import audioWfc.wfc.hierarchy.prototypes.ChordPrototype;
import audioWfc.wfc.hierarchy.prototypes.Chordesque;

import java.util.Set;

public class ChordPrototypeOnlyPrecededByConstraint implements HardConstraint<Chordesque> {
    private Grabber<Set<Chord>> precedingChordsGrabber;
    private String chordPrototypeName;

    public ChordPrototypeOnlyPrecededByConstraint(Grabber<Set<Chord>> precedingChordsGrabber, String chordPrototypeName) {
        this.precedingChordsGrabber = precedingChordsGrabber;
        this.chordPrototypeName = chordPrototypeName;
    }

    @Override
    public String name() {
        return "Section is followed by";
    }

    @Override
    public String configText() {
        return precedingChordsGrabber.configText();
    }

    @Override
    public boolean check(Tile<Chordesque> tile, HigherValues higherValues) {
        Chordesque chordesque = tile.getValue();
        boolean out = true;

        Tile<Chordesque> prevTile = tile.getPrev();
        if(prevTile.isCollapsed()){
            out &= check(prevTile.getValue(), chordesque, higherValues);
        }

        Tile<Chordesque> nextTile = tile.getNext();
        if(prevTile.isCollapsed()){
            out &= check(chordesque, nextTile.getValue(), higherValues);
        }

        return out;
    }

    private boolean check(Chordesque a, Chordesque b, HigherValues higherValues){
        if(!(b instanceof ChordPrototype c))return true;
        if (!c.getName().equals(chordPrototypeName)) return true;
        Set<Chord> precedingChords = precedingChordsGrabber.grab(higherValues);
        return precedingChords.contains(a.getValue());
    }
}

File: vite-project/src/wfc/constraints/ChordPrototypeOnlyPrecededByConstraint.ts
import { Chord } from "../../music_theory/Chord"
import { HigherValues } from "../HigherValues"
import { Tile } from "../Tile"
import { Chordesque, ChordPrototype } from "../hierarchy/prototypes"

export class ChordPrototypeOnlyPrecededByConstraint implements HardConstraint<Chordesque> {
	private chordPrototypeName: string
	private grabber: Grabber<Set<Chord>>

	name = "Chord Prototype Only Preceded By"
	configText = () => `Chord Prototype: ${this.chordPrototypeName}, Chord Set: ${this.grabber.configText()}`

	constructor(chordPrototypeName: string, grabber: Grabber<Set<Chord>>) {
		this.chordPrototypeName = chordPrototypeName
		this.grabber = grabber
	}

	check(tile: Tile<Chordesque>, higherValues: HigherValues): boolean {
		const chordesque = tile.getValue()
		const prev = tile.getPrev()
		const next = tile.getNext()

		let out = true
		if(prev.isCollapsed()) out = out && this.checkPair(prev.getValue(), chordesque, higherValues)
		if(next.isCollapsed()) out = out && this.checkPair(chordesque, next.getValue(), higherValues)
		return out
	}

	private checkPair(first: Chordesque, second: Chordesque, higherValues: HigherValues): boolean {
		if(!(second instanceof ChordPrototype)) return true
		if(second.getName() !== this.chordPrototypeName) return true
		const chordSet = this.grabber.grab(higherValues)
		return chordSet.has(first.getChord())
	}
}